### DISCUSSION TITLE
Use for browser extensions

### DISCUSSION THREAD

**Discussion by marcelreppi** [umami-software/umami#837] [General]
I wanted to use umami to track events in a browser extension, however I can't make it work for some reason.
~~Using the `script` tag doesn't work due to CSP~~ I got it working only for the Vue.js files of my browser extension but not for any background scripts. I tried downloaded the script and import it instead and do the tracking via JS but `umami` is not added to the `window` object for the background scripts.

Any ideas how to get this working for browser extensions? 

Thanks!

**Comment by marcelreppi** [comment]
I now got it working just for the Vue.js files of my browser extension by setting the `content_security_policy` in the `manifest.json` and using the `script` tag as usual in my `index.html`.
However I still have an issue using it in background scripts. `umami` is not available on the `window` object.

EDIT: I guess the issue is this line since I import the tracker script via my `manifest.json` and there is no `script` tag for background scripts.
https://github.com/mikecao/umami/blob/2575cbfc11c0c59c3b241cab51bad2f9e1b0b62f/tracker/index.js#L15

**Comment by marcelreppi** [comment]
I got it working for background scripts by using a background page instead of single scripts but that kind of breaks my application.
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background
```
  "background": {
    "page": "my-background.html"
  }
  ```

I would highly prefer using the `scripts` syntax 😞 
```
  "background": {
    "scripts": ["umami.js", "my-background.js"]
  }
```

**Comment by marcelreppi** [comment]
Got it working now with the `script` syntax by altering the downloaded `umami.js` file the following way:
1. `m = v.querySelector("script[data-website-id]")` to `m = v.querySelector("script[src*='umami']")`
2. `w = g("data-website-id")` to `w = "WEBSITE_ID"`
3. `y = g("data-host-url")` to `y = "HOST_URL"`

Hope this helps anyone running into the same issue.

Maybe you could make it more easily usable for browser extensions in the future :)

**Reply by yonghui666** [reply]
I have a question. It's been 5 years now. Are there any more suitable statistics solutions for browser extensions?

**Reply by marcelreppi** [reply]
No idea. I haven't looked into it further since I didn't have any additional need.

**Comment by mikecao** [comment]
Not sure what a browser extension would be for, but moving this to discussions since it's not an issue.

**Reply by rarepepi** [reply]
can we get an update on this, im trying to use umami for browser extension @marcelreppi thanks!

**Reply by yonghui666** [reply]
Can you provide the latest progress? I am trying to use Umami as a browser extension.

**Comment by BlaineMoore** [comment]
Thanks @marcelreppi for figuring this out! I'm currently testing Umami and Plausible to see which will work better for me as we rewrite our browser extension, so being able to track page views and events in our extension is going to be a priority. Good to see that there may be a way to get this to work out of the box, as it were...
